File: Binary Search/367. Valid Perfect Square/367. Valid Perfect Square.py
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        l = 0
        r = num
        while l <= r:
            m = (l + r) // 2
            s = m * m
            if s == num:
                return True
            elif s < num:
                l = m + 1
            else:
                r = m - 1
        return False

# Time: O(log n)
# Space: O(1)

File: Binary Search/367. Valid Perfect Square/367. Valid Perfect Square.js
/**
 * @param {number} num
 * @return {boolean}
 */
var isPerfectSquare = function (num) {
  let l = 0;
  let r = num;
  while (l <= r) {
    const m = Math.floor((l + r) / 2);
    const s = m * m;
    if (s === num) return true;
    else if (s > num) r = m - 1;
    else l = m + 1;
  }
  return false;
};

// Time: O(log n)
// Space: O(1)
